doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+pandas-specific metadata objects that model the current semantics / behavior of
+the project. What does this mean, exactly?
+
+* Each NumPy dtype object will map 1-to-1 to an equivalent ``pandas.DataType``
+  object.
+* Existing pandas "extension dtypes" (like ``CategoricalDtype`` and
+  ``DatetimeTZDtype``), which have been designed to mimic ``numpy.dtype``, will
+  become logical type subclasses of ``pandas.DataType`` like every other type
+  in pandas.
+
+Since pandas is about assisting with data manipulation and analysis, at some
+point you must invoke functions that are specialized to the specific physical
+memory representation of your data. For example, pandas has its own
+implementations of ``ndarray.take`` that are used internally for arrays of
+positive integers that may contain NULL / NA values (which are represented as
+-1 -- search the codebase for implementations of ``take_1d``).
+
+The major goals of introducing a logical type abstraction are the follows:
+
+* Simplifying "dynamic dispatch": invoking the right functions or choosing the
+  right code branches based on the data type.
+* Enabling pandas to decouple both its internal semantics and physical storage
+  from NumPy's metadata and APIs. Note that this is already happening with
+  categorical types, since a particular instance of ``CategoricalDtype`` may
+  physically be stored in one of 4 NumPy data types.
+
+Physical storage decoupling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By separating pandas data from the presumption of using a particular physical
+``numpy.dtype`` internally, we can:
+
+* Begin to better protect users from NumPy data semantics (which are frequently
+  different from pandas's!) leaking through to the pandas user API. This can
+  enable us to address long-standing inconsistencies or "rough edges" in pandas
+  that have persisted due to our tight semantic coupling to NumPy.
+
+* We can consider adding new data structures to pandas, either custom to pandas
+  or provided by 3rd-party libraries, that add new functionality alongside the
+  existing code (presuming NumPy physical storage). As one concrete example,
+  discussed in more detail below, we can enable missing data in integer pandas
+  data by forming a composite data structure consisting of a NumPy array plus a
+  bitmap marking the null / not-null values.
+
+* We can start to think about improved behavior around data ownership (like
+  copy-on-write) which may yield many benefits. I will write a dedicated
+  section about this.
+
+Note that neither of these points implies that we are trying to use NumPy
+less. We already have large amounts of code that implement algorithms similar
+to those found in NumPy (e.g. ``pandas.unique`` or the implementation of
+``Series.sum``), but taking into account pandas's missing data representation,
+etc. Internally, we can use NumPy when its computational semantics match those
+we've chosen for pandas, and elsewhere we can invoke pandas-specific code.
+
```

Comment:
I very much agree with this and think it is a crucial point that could be emphasised more in the docs.  I suspect that this is a tripping point for many users regarding the application of `numpy` functionality on `pandas` objects.